Refactor ofxGui with extended features and controls

frauzufall opened the pull request:
I am creating this pull request to have a starting point for the discussion about this.

frauzufall commented:
You cleared out a lot of inherited functions in your PR, we should do the same here. 

This PR also has unresolved layout issues - the positioning and sizing is not that good. I was not sure how spacingNextElement and spacingFirstElement were used and how to apply that to the vertical layout. There were also issues of ofxGuiGroup in the core addon when the container gets resized before or after adding controls, that's why I could not always keep the original positioning.

We should apply your PR and the newer commits to this branch with this PR and then ping the people who wanted to help - roymacdonald and thomasgeissl - to do testing and discuss the rest? Like if we keep the rotary slider and the graphics and things like that. I also would have suggested to put the master slave controls into a separate addon. But I could not yet get them to work with the new ofxGui so it could still affect this branch.

And then there are some additional suggestions - slider controlled by arrow keys (https://github.com/roymacdonald/ofxGuiPlus) and the wish to disable controls. This would be another variable in ofxBaseGui.

arturoc commented on addons/ofxGui/src/ofxBaseGui.cpp:
```diff
@@ -225,22 +227,28 @@ void ofxBaseGui::setPosition(float x, float y){
 
-void ofxBaseGui::setSize(float w, float h){
+void ofxBaseGui::setSize(float w, float h, bool callback){
 	b.width = w;
```
let's remove this callbacks parameters, it's a pattern that we use sometimes in OF and it's really confusing to read.

this:

```
gui.sizeChangedE.disable()
gui.setSize(40,40)
gui.sizeChangedE.enable()
```

is more verbose but it's clear what it means, but this:

```
gui.setSize(40,40,false);
```
it's really cryptic and forces you to go to the original headers to see the meaning of that parameter, another possibility would be something like:

```
gui.setSizeWithoutEvent(40,40);
```

but it would need to be public and since it's something we only need internally to avoid recursive calls i think the first syntax is ok

arturoc commented:
also not completely sure but we might want to join the sliders in one class only, i added a Layout enum and variable in ofxBaseGui, we could switch on that to generate the slider as vertical or horizontal instead of having 2 classes. then a horizontal gui group will generate sliders with vertical layout and a vertical gui group horizontal sliders by default but we can configure that through the config. 

other controls could switch on that too to render differently depending on the layout.

what do you think?

in any case i think adding back that layout variable and enum in the base class and removing bVertical from ofxGuiGroup is more flexible

frauzufall commented:
Yes all these suggestions make sense to me. Who wants to rewrite? I can do it, just making sure we are not doing the same thing again :-) 

Any idea on how to set the child's config to the config of the parent but also keeping the ability to set certain values in the child's constructor? (see last commit)

arturoc commented:
go for it, let me know if you need help with something.

i saw your latest commit before, it's definitely better than before, i thought about removing it but finally forgot. the config in the guigroup doesn't make much sense, i added only to do this and it's not even kept up to date internally. the only things i finally needed to set up were the layout and a flag to tell the created gui that it lives inside a container since some gui elements render slightly different depending if they are contained or not:

https://github.com/arturoc/openFrameworks/blob/refactor-gui/addons/ofxGui/src/ofxGuiGroup.h#L220

if you can add back those 2 flags in ofxBaseGui and set them to children in those methods in ofxGuiGroup i think that should be enough

frauzufall commented:
Do I have to do this: 
```
ofxFpsPlotter::ofxFpsPlotter(const ofxValuePlotter::Config & c)
```
I should be able to just use `Config & c` or did you add the class for better readability?

frauzufall commented:
and while I am at this point - why did you remove the empty constructor and setup function of ofxValuePlotter? To force one to use a parameter with the plotted value? I like the idea of using only the parameter, I just ask if we shouldn't keep the other option. We have this approach:
```c++
//.h
ofParameter<float> value;
//.cpp setup
add<ofxValuePlotter>(value);
//.cpp update
value.set(0.5);
```
and this one:
```c++
//.h
ofxValuePlotter plotter;
//.cpp setup
add(plotter.setup());
//.cpp update
plotter.update(0.5)
```
You think we should remove the second one? I am fine with that, I just want to clarify.

arturoc commented:
i'm not 100% sure but i think we should deprecate the setup option and having gui elements directly in user code in favor of creating parameters and adding them to a gui group. i've kept the older setup methods so old code keeps working but though that for new elements we could just remove them from the beginning. On the other hand if this is going to be part of ofxGui it will be kind of not very uniform so not very sure.

if you already adapted the setup methods to return a reference let's leave them in by now

arturoc commented:
```
ofxFpsPlotter::ofxFpsPlotter(const ofxValuePlotter::Config & c)
```

because i'm using the config struct from the parent class, thinking of it we could add a config struct in this one too even if it's empty just for uniformity sake

frauzufall commented:
Alright sorry. I didn't read the fps plotter constructor carefully.

I like the idea of making ofxGui solely parameter based. I would leave the setup functions in and mark them all deprecated. But what about buttons? Should the button-pressed-event also be changed to a parameter-changed-event? That's probably not the most intuitive way?

frauzufall commented:
Another question: Why don't the groups get configs and config constructors? Couldn't they also be used without a global variable or is the `getControl`-function going to be gone because it can be misused?
```c++
ofxGuiGroup::Config config;
config.alignment = OfxBaseGui::Horizontal;
add<ofxGuiGroup>("mygroup",config);
getControl("mygroup")->add(parameter1);
getControl("mygroup")->add(parameter2);
getControl("mygroup")->add(parameter3);
```
I am not completely happy with this because there is no check if the names are unique and there can be typos without the compiler noticing etc. but I like to be able to get rid of the gobal gui control variables. What do you think?

arturoc commented:
yes i was doubting on groups because they are the only case where you probably want to use them after creation and i don't like the getControl() method much either. it's using pointers which i'm trying to avoid, it can fail without compiler warning as you say and in that case it'll return null and crash the program.

I'll give this a try at some point but let's leave them as they are by now. Having a config object though makes sense but keeping the setup methods so they can be initialized with the old style makes sense by now too.

In general i think the guidelines might be:
- avoid pointers, the syntax is strange for beginners but also the ownership is never clear ( we could use shared_ptr though)
- most controls don't need to be modified after initialization so the config object allows to pass any parameters from the beginning
- make the syntax as easy to use as possible.

frauzufall commented:
Yes, I keep that in mind.
One minor thing - why do we need the two generateText-functions with identical content in the slider class? What do we need the one with the unsigned char type for?
```
template<typename Type>
void ofxSlider<Type>::generateText()

template<>
void ofxSlider<unsigned char>::generateText()
```

frauzufall commented:
Adapted to everything we discussed.

I just had to comment out that if you add a control to a group it gets the groups layout. This is too inconsistent at the moment. When the group has a horizontal layout, the sliders would be vertical, but their layout attribute would be horizontal. When you then want to change the sliders to be horizontal, you would need to set the layout variable to vertical. The other inconsistent thing is that all the other controls like the toggles don't change in a horizontal group. That's why I set all controls to horizontal and the groups to vertical by default and you have to change the controls of a group too if you want a different layout for now.

frauzufall commented:
I also removed ofxGuiGraphics and ofxGuiZoomableGraphics to keep it to parameter controlled interfaces.

arturoc commented:
this:

```
template<typename Type>
void ofxSlider<Type>::generateText()

template<>
void ofxSlider<unsigned char>::generateText()
```

specializes the templated function to work different for char to be cast to int when passed through the << operator, otherwise they are printed as characters instead of numbers. you can see the problem if you remove that function and add a color slider which would print as strange characters instead of numbers.

> I just had to comment out that if you add a control to a group it gets the groups layout. 

yes i thought about that but didn't really tested it, makes total sense, even the size that would be passed to the children won't make sense. just passing the correct layout and setting the inContainer flag in children should be enough.

> I also removed ofxGuiGraphics and ofxGuiZoomableGraphics to keep it to parameter controlled interfaces. 

ok perfect. i'm not sure that all the guis should have a parameter, something like a spacer won't have mouse interaction or a parameter but it's still a gui element, we might want to refactor the base class into 2 or 3 abstract classes at some point but lets leave it as it is by now. i'm not sure about the utility of ofxGuiGraphics any way so better to keep them out by now

thanks for all the work on this!!

frauzufall commented:
What you described about the color slider labels and the `ofxSlider<unsigned char>::generateText()` function is happening right now (the strange chars) even though I have both methods in ofxSlider.cpp. Do you have an idea what else I could have done that leads to this error?

I use something like the ofxGuiGraphics class a lot. I have an ofFbo that is created in multiple steps using multiple other ofFbos that I can modify and I need ofxGuiGraphics to show the intermediate steps within the gui (like in <a href="http://forum.openframeworks.cc/uploads/default/original/2X/a/a66301d7d2bfc03bb6e596e4ddd6aa0375ed855f.png">this screenshot</a> I posted in the forum). It could also be used for example to show the camera input as a (zoomable) preview object in the gui. But I am completely fine with keeping them in ofxGuiExtended.

frauzufall commented:
I got a problem: I want to update my ofxSortableList (https://github.com/frauzufall/ofxSortableList) to the new style. I can't, because I can't swap elements within `collection` of `ofxGuiGroup`. Can we add this? It's not completely trivial because one also has to swap the parameter in the parameter group and this is also not implemented.

arturoc commented:
didn't saw that before, it's really nice! :)   

i guess we can add methods to ofParameterGroup that allow to do that, another option is to just recreate the full parameter group with the new order which shouldn't be a problem in terms of performace since it's something that won't happen that often.

frauzufall commented:
That worked well, thank you! I could update ofxSortableList without changing ofxGui.

The bug with the strange characters when using ofxColorSlider remains..

Another thing that seems unreasonable is that a button with a listener calls the listener function when it is added to a group.

frauzufall commented:
I fixed the color slider char bug (it was actually already fixed in your refactor-gui branch) and the parameters that got triggered when added to the gui were my fault (I added the listener before adding them to the gui). 

Some small changes:
- `ofxLabel` is now a derivative from `ofxValueLabel` that is able to display other `ofParameter` values besides `string`. `ofxLabel` can be used the same way as before. To display other data types, use `ofxValueLabel<Type>` or `ofxIntLabel`, `ofxFloatLabel` or `ofxBoolLabel`. 
- ofxSlider has a precision option (not applied to the value, only to the displayed value)
- I wanted to update to the current master and now I have all these unrelated commits in this PR -.-

You are probably busy with 0.9 so I will just post this branch to the forum to get some testing done. I am also using this branch to refactor my apps so I already use it a lot.

I also want to add the feature to disable controls, apply its own font size to a control (right now font changes affect all controls?) and think about the feature to control sliders with arrow keys or direct input as suggested <a href="http://forum.openframeworks.cc/t/extending-ofxgui-with-new-widgets/16670/23?u=frauzufall">in the forum</a>.

arturoc commented:
that spunds great, yes i've been crazy busy with a project and 0.9 but will take a look as soon as i have a moment
